### ISSUE TITLE
[CI] Setup Bazel Remote Cache

### ISSUE THREAD

**Issue by simon-mo** [ray-project/ray#4982]
[Bazel has the ability](https://docs.bazel.build/versions/master/remote-caching.html) to cache its output to a remote server. This can greatly speed up our CI build and other build. 

We can setup a file server that gives:
- read+write access to Travis CI builds
- read access to anyone else using and developing ray. 

I was able to setup a proof of concept server and saw 10x speed up build ray on my laptop (10min to 46s)! 

TODO for PR (#PLACEHOLDER):
- [ ] Discuss who and how to host the file server
- [ ] Make sure the server hosting is reproducible (commit nginx config to PR)
- [ ] Setup permission correctly

**Comment by raulchen** [comment]
Cool, I was also playing with bazel remote cache today. I came across [this probject](https://github.com/Asana/bazels3cache) that allows using S3 as remote cache backend.  Did you try this?

**Comment by simon-mo** [comment]
@raulchen Can I ask what would be the advantage of using s3+proxy over the simple nginx server approach? 

ray can host a file server and enable everyone to read from it (they don't need to depend on nodejs). Anyone can host the server for their own build as well. 

To deploy such a webserver, it's as simple as getting a machine with nginx enabled and add this:
```nginx
server {
    server_name ray-bazel.simon-mo.com;
    client_body_temp_path  /tmp;
    autoindex on;
    location / {
	  # The path to the directory where nginx should store the cache contents.
	  root /home/ubuntu/server-files;
	  # Allow PUT
	  dav_methods PUT;
	  # Allow nginx to create the /ac and /cas subdirectories.
	  create_full_put_path on;
	  # The maximum size of a single file.
	  client_max_body_size 1G;
	  allow all;

          # Password
	    auth_basic              ci_access;
	    auth_basic_user_file    /etc/nginx/.passwords.list;
    }
}
```

**Comment by simon-mo** [comment]
if the concern is scalability, a single node cache server can handle quite a lot of requests already. We can also just delete the old cached files when the disk is getting full

**Comment by raulchen** [comment]
@simon-mo 
I was mainly thinking about speeding up CI builds. As introducing gRPC (https://github.com/ray-project/ray/pull/4968) adds 10+ minutes build time. And using local proxy + S3 seems the quickest solution.

If we have enough resources to host the cache server for the whole community, that'd be definitely wonderful. BTW, even if we have such a server, we should still make the server address configurable, because users may want to host their own server for fast network connection.

**Comment by pcmoritz** [comment]
I think we should try it with a webserver and a configurable URL (plus instructions to set up the server for people who want to use their own). The number of people who are building Ray is probably not super large (most people install the wheels), so a single webserver that we provide might be good enough :)

**Comment by simon-mo** [comment]
Closed by #5068 and #5144
